File: Data Structures & Algorithms/linked-list-cycle-detection/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        ListNode* cur1 = head;
        ListNode* cur2 = head;
        while (cur2) {
            cur1 = cur1->next;
            cur2 = cur2->next;
            if (not cur2){
                return false;
            }
            cur2 = cur2->next;
            if (cur1 == cur2){
                return true;
            }
        }
        return false;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        cur = head
        end = head
        if not head.next:
            return
        while cur:
            end = cur
            while end.next:
                prev = end
                end = end.next
            prev.next = None
            temp = cur.next    
            cur.next = end
            end.next = temp
            cur = cur.next
            if not cur:
                break
            cur = cur.next